File: Week11/P2178.py
from collections import deque

def shortest_path(N,M,grid):
    #방향 벡터 (상,하 좌,우)
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]

    #방문여부 저장배열 
    visited = [[False for _ in range(M)]for _ in range(N)]

    #BFS 큐 초기화 (x, y, 거리)
    queue = deque([(0,0,1)])
    #시작점 방문여부 처리하기 
    visited[0][0] = True

    while(queue):

        #먼저 들어간 좌표부터 꺼내기 
        x,y,dist = queue.popleft()

        #목적지에 도달하였을 때, 
        if x == N-1 and y == M-1:   
            return dist
        
        #4가지 방향 탐색 
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            #좌표가 범위 안에 들어있고, 방문한 적이 없을 때 
            if (0 <= nx < N and 0 <= ny < M and visited[nx][ny] == False):
                #이동할 수 있는 칸일 경우 
                if(grid[nx][ny] == 1):
                    queue.append((nx,ny,dist+1))
                    #방문 여부 기록
                    visited[nx][ny] = True

#N은 행, M은 열
N, M = map(int, input().split())

#map 저장 -> 이차원 배열 저장 
grid = [list(map(int, input().rstrip()))for _ in range(N)]

result = shortest_path(N, M, grid)
print(result)

File: Week11/P2178.java

import java.util.*;

public class P2178{
    //x, y 좌표와 이동거리 저장 Class
    static class Node{
        int x, y, dist;

        Node(int x, int y, int dist){
            this.x = x;
            this.y = y;
            this.dist = dist;
        }
    }

    public static int shortest_path(int n, int m, int [][] grid){

        //상, 하, 좌, 우 
        //배열 한번에 초기화 할때 중괄호 사용
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        //방문 여부 기록 배열 
        boolean [][] visited = new boolean[n][m];
        Queue<Node> queue = new LinkedList<>();

        //큐와 방문여뷰 배열에 시작 값 넣기 
        queue.offer(new Node(0,0,1));

        visited[0][0] = true;

        while(!queue.isEmpty()){

            //현재 노드 꺼내기 
            Node current = queue.poll();
            int x = current.x;
            int y = current.y;
            int dist = current.dist;

            //정착지에 도달했을 경우 
            if(x == n-1 && y == m-1){
                return dist;
            }
            //상,하,좌,우 탐색
            for (int i = 0; i < 4; i++) {

                int nx = x + dx[i];
                int ny = y + dy[i];

                //범위 안에 있는 경우    
                if (0 <= nx && nx < n && 0 <= ny && ny < m){
                    //방문한 적이 없고, grid 값이 1인 경우 
                    if(visited[nx][ny] == false && grid[nx][ny]==1){
                        //만족할 경우 큐에 추가 
                        queue.offer(new Node(nx, ny, dist+1));
                        //방문 여부 기록
                        visited[nx][ny] = true;
                    }

                }

            }

        }
        return -1;

    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // N 행, M 열 저장
        int N = sc.nextInt();
        int M = sc.nextInt();

        //gri 저장 
        int [][] grid = new int[N][M];

        for (int i = 0; i < N; i++){
            String row = sc.next(); //공백 전까지 한 줄 입력받기 
            for(int j = 0; j <M; j++){
                grid[i][j] = row.charAt(j) -'0';//문자 정수로 변환
            }
        }

        int result = shortest_path(N, M, grid);
        System.out.println(result);
    }
}